subsys/bluetooth/CMakeLists.txt

```diff
@@ -1,10 +1,20 @@
-add_library(subsys__bluetooth INTERFACE)
+# We are just defining avaiable libaries created in the Bluetooth subsystem in order to provide
+# easy understanding og what is available Mesh is part of host in this example.
 
-target_include_directories(subsys__bluetooth INTERFACE ${CMAKE_CURRENT_SOURCE_DIR})
+add_library(subsys_bluetooth STATIC "")
 
-add_subdirectory(common)
-add_subdirectory_ifdef(CONFIG_BT_HCI host)
-add_subdirectory_ifdef(CONFIG_BT_CTLR controller)
-add_subdirectory_ifdef(CONFIG_BT_SHELL shell)
+include(${CMAKE_CURRENT_LIST_DIR}/common/bluetooth_common.cmake)
```

Comment:
This also breaks every other use-case that is covered by the existence of ZEPHYR_LIBS_PROPERTY.

So what is lacking from this RFC is an overview of all the cases where a global list of libraries is used and new strategies for each of them.

Reply:
ZEPHYR_LIBS_PROPERTY is only used in <zephyr-repo>/CMakeLists.txt file.

And it is used for three things, as I can see:
1) Looping the libs to ensure they are not empty:
https://github.com/zephyrproject-rtos/zephyr/blob/master/CMakeLists.txt#L499-L508
and if the libray is imported, then it is disregarded.
Thus, only libraries created by Zephyr's build system is checked.
In CMake, if you specify a libray and don't add sources, then you get the following error:
'No SOURCES given to target: <library>'
(except for INTERFACE libs, but those are not checked here)
So this RFC should also remove the need for this checked.
Not sure why the check is here, but could imagine that using a list with all libraries, is the root cause here, as someone could add invalid targets to the list
2) at set(zephyr_lnk ...)
https://github.com/zephyrproject-rtos/zephyr/blob/master/CMakeLists.txt#L706-L713
Which will be obsolete with this RFC
3) Kernel / userspace
https://github.com/zephyrproject-rtos/zephyr/blob/master/CMakeLists.txt#L526
This is a real problem.
Will look into a proposal, and update the RFC with that.

Reply:
About the empty check. It is OK to drop this check, it was added out of convenience since the list already existed.

The CMake version that I was using at the time was giving an obscure error message about the wrong language of source files being used and nothing about the library being empty, which is why the check was introduced.

Reply:
I have added a change with improved handling of kernel / app memory space location.

A custom target memory_space_kernel is added.
The memory_space_kernel is then populated with those targets, such as kernel, subsys_bluetooth, subsys_net, etc.

Having much fewer libraries makes it less cumbersome to specify which libraries should be in kernel space explicitly, and much clearer to the developer.
Furthermore it gives developers using Zephyr to place their own libs in kernel space.

A general function has been introduced, which follows same style as CMake set_property but handles specific Zephyr properties, currently MEMORY_SPACE kernel, but may be extended in future.
Example:
zephyr_set_property(TARGET subsys_net MEMORY_SPACE kernel)

Advantage of this approach over current:
1) Libs are not place in kernel space based on guessing from their location / name
2) Libs are explicitly stated when they are kernel space candidate
3) Using of target (over global variable) ensures that spelling mistakes are causing an error, eg. writing  zephyr_set_property(TARGET subsys_nett MEMORY_SPACE kernel) results in:   get_target_property() called with non-existent target "subsys_nett".
Call Stack (most recent call first):
  ../../../subsys/net/CMakeLists.txt:25 (zephyr_set_property)
4) Using target and target properties that supports generator expressions removes the configure time dependency of CMake code ordering. i.e zephyr_set_property() can be used in anyplace of cmake code, whereas previously any calls that populating ${ZEPHYR_LIBS_PROPERTY} had to be placed before: https://github.com/zephyrproject-rtos/zephyr/blob/8a74e7056dc6e405b1336e047957d143af5b72d9/CMakeLists.txt#L534-L625


Disadvantage:
1) Having to write zephyr_set_property(TARGET subsys_net MEMORY_SPACE kernel). However, with the reduced number of libraries that this RFC propose, this becomes a much smaller issue, and I think the pros out weights the cons.